### PR TITLE
Fix thread safety issues with buffer hash

### LINKED ISSUE #12: BufferedLogger::AlreadyStartedError errors happening in Puma workers
Not too sure what could of cause this but started seeing allot of these errors. Had to restart puma workers. 

```
2015-09-30_02:05:19.30691 2015-09-30 02:05:19 +0000: Rack app error: #<BufferedLogger::AlreadyStartedError: already started>
2015-09-30_02:05:19.30697 /shared/bundle/ruby/2.1.0/gems/buffered-logger-1.2.2/lib/buffered_logger.rb:29:in `start'
2015-09-30_02:05:19.30700 /shared/bundle/ruby/2.1.0/gems/buffered-logger-1.2.2/lib/buffered_logger/middleware.rb:8:in `call'
2015-09-30_02:05:19.30702 /shared/bundle/ruby/2.1.0/gems/airbrake-4.0.0/lib/airbrake/user_informer.rb:16:in `_call'
2015-09-30_02:05:19.30705 /shared/bundle/ruby/2.1.0/gems/airbrake-4.0.0/lib/airbrake/user_informer.rb:12:in `call'
2015-09-30_02:05:19.30707 /shared/bundle/ruby/2.1.0/gems/railties-4.1.11/lib/rails/engine.rb:514:in `call'
2015-09-30_02:05:19.30710 /shared/bundle/ruby/2.1.0/gems/railties-4.1.11/lib/rails/application.rb:144:in `call'
2015-09-30_02:05:19.30714 /shared/bundle/ruby/2.1.0/gems/puma-2.11.1/lib/puma/configuration.rb:82:in `call'
2015-09-30_02:05:19.30716 /shared/bundle/ruby/2.1.0/gems/puma-2.11.1/lib/puma/server.rb:507:in `handle_request'
2015-09-30_02:05:19.30719 /shared/bundle/ruby/2.1.0/gems/puma-2.11.1/lib/puma/server.rb:375:in `process_client'
2015-09-30_02:05:19.30722 /shared/bundle/ruby/2.1.0/gems/puma-2.11.1/lib/puma/server.rb:262:in `block in run'
2015-09-30_02:05:19.30724 /shared/bundle/ruby/2.1.0/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `call'
2015-09-30_02:05:19.30726 /shared/bundle/ruby/2.1.0/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `block in spawn_thread'
```

My guess is this could be a thread safety issue but not sure what conditions might trigger this.

### PR DESCRIPTION
**Pull request by sunblaze**
**TLDR** Rewrote LogDeviceProxy to use thread local variables and hopefully make the gem truly thread safe.

Fixes #12 

Added two slow tests that through brute force can create the following exceptions with the old implementation of a buffer hash. 

```
BufferedLogger::LogDeviceProxy::flush#test_0001_should handle multiple thread simutaneously calling it:
NoMethodError: undefined method `string' for nil:NilClass
    /Users/james/Code/buffered-logger/lib/buffered_logger/log_device_proxy.rb:19:in `flush'
    /Users/james/Code/buffered-logger/test/buffered_logger/log_device_proxy_test.rb:99:in `block (7 levels) in <top (required)>'
    /Users/james/Code/buffered-logger/test/buffered_logger/log_device_proxy_test.rb:97:in `times'
    /Users/james/Code/buffered-logger/test/buffered_logger/log_device_proxy_test.rb:97:in `block (6 levels) in <top (required)>'
```

```
BufferedLogger#test_0003_should handle multiple thread simutaneously calling start with a block:
BufferedLogger::AlreadyStartedError: already started
    /Users/james/Code/buffered-logger/lib/buffered_logger.rb:29:in `start'
    test/buffered_logger_test.rb:36:in `block (6 levels) in <main>'
    test/buffered_logger_test.rb:35:in `times'
    test/buffered_logger_test.rb:35:in `block (5 levels) in <main>'
```

```
BufferedLogger#test_0003_should handle multiple thread simutaneously calling start with a block:
BufferedLogger::NotStartedError: not started
    /Users/james/Code/buffered-logger/lib/buffered_logger.rb:18:in `end'
    /Users/james/Code/buffered-logger/lib/buffered_logger.rb:36:in `start'
    test/buffered_logger_test.rb:36:in `block (6 levels) in <main>'
    test/buffered_logger_test.rb:35:in `times'
    test/buffered_logger_test.rb:35:in `block (5 levels) in <main>'
```

```
BufferedLogger#test_0003_should handle multiple thread simutaneously calling start with a block:
NoMethodError: undefined method `string' for nil:NilClass
    /Users/james/Code/buffered-logger/lib/buffered_logger/log_device_proxy.rb:15:in `end'
    /Users/james/Code/buffered-logger/lib/buffered_logger.rb:19:in `end'
    /Users/james/Code/buffered-logger/lib/buffered_logger.rb:36:in `start'
    test/buffered_logger_test.rb:36:in `block (6 levels) in <main>'
    test/buffered_logger_test.rb:35:in `times'
    test/buffered_logger_test.rb:35:in `block (5 levels) in <main>'
```

Or sometimes would seg fault MRI :worried: 

```
/Users/james/Code/buffered-logger/lib/buffered_logger/log_device_proxy.rb:15: [BUG] Segmentation fault at 0x007f8f44100000
ruby 2.1.7p400 (2015-08-18 revision 51632) [x86_64-darwin14.0]

-- Crash Report log information --------------------------------------------
   See Crash Report log file under the one of following:
     * ~/Library/Logs/CrashReporter
     * /Library/Logs/CrashReporter
     * ~/Library/Logs/DiagnosticReports
     * /Library/Logs/DiagnosticReports
   for more details.

-- Control frame information -----------------------------------------------
c:0008 p:---- s:0025 e:000024 CFUNC  :delete
c:0007 p:0015 s:0021 e:000019 METHOD /Users/james/Code/buffered-logger/lib/buffered_logger/log_device_proxy.rb:15
c:0006 p:0029 s:0017 e:000016 METHOD /Users/james/Code/buffered-logger/lib/buffered_logger.rb:19
c:0005 p:0048 s:0014 e:000012 METHOD /Users/james/Code/buffered-logger/lib/buffered_logger.rb:36
c:0004 p:0009 s:0009 E:001e68 BLOCK  test/buffered_logger_test.rb:36 [FINISH]
c:0003 p:---- s:0007 e:000006 CFUNC  :times
c:0002 p:0010 s:0004 E:002128 BLOCK  test/buffered_logger_test.rb:35 [FINISH]
c:0001 p:---- s:0002 e:000001 TOP    [FINISH]

test/buffered_logger_test.rb:35:in `block (5 levels) in <main>'
test/buffered_logger_test.rb:35:in `times'
```

All the exceptions related to simultaneously accessing the @buffers hash in LogDeviceProxy. I didn't see a reason why we couldn't move the current StringIO to a thread local variable so I re-implemented it using thread local. Otherwise we'd need a mutex around access of the @buffers hash or more of the dup the hash dance we were doing before.

Bumped the version number along with these changes and also deprecated the sweep functionality since it was no longer needed with the new implementation.

### CODE REVIEW

**Comment by garethson** [comment]
Ping @samuelkadolph. We'd like to use this for something using Puma, thoughts?

File: lib/buffered_logger.rb [hunk 1]
```diff
@@ -45,7 +41,17 @@ def started?
   end
 
   def sweep
-    @logdev.sweep
+    warn "[DEPRECATION] `sweep` is deprecated.  It is no longer necessary."
```
Comment (garethson) [inline lib/buffered_logger.rb]: I think that if we're bumping this a major version here, we don't really need to deprecate, we can just delete these.  Thoughts?

**Comment by garethson** [comment]
This is great. Thanks!

File: lib/buffered_logger.rb (hunk 1 above)
Comment (andremedeiros) [inline lib/buffered_logger.rb]: Totally agree. Scrub it, especially because it doesn't even do what you want it to do anymore.

File: lib/buffered_logger/log_device_proxy.rb [hunk 2]
```diff
@@ -12,44 +14,54 @@ def close
...
-      @buffers[key].string.dup
+      string_io.string.dup
     end
 
     private
-      def key
-        [Thread.current]
+    def string_io
+      if Thread.current.respond_to?(:thread_variable_get)
+        Thread.current.thread_variable_get(THREAD_LOCAL_VAR_NAME)
```
Comment (samuelkadolph) [inline lib/buffered_logger/log_device_proxy.rb]: Why use this instead of just `#[]`?

File: lib/buffered_logger.rb (hunk 1 above)
Comment (kmcphillips) [inline lib/buffered_logger.rb]: +1

File: test/buffered_logger/log_device_proxy_slow_test.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require "test_helper"
+
+describe "BufferedLogger::LogDeviceProxy slow tests" do
+  before do
+    @logdev = mock()
+    @proxy = BufferedLogger::LogDeviceProxy.new(@logdev)
+  end
+
+  describe "flush" do
+    it "should handle multiple thread simutaneously calling it" do
```
Comment (kmcphillips) [inline test/buffered_logger/log_device_proxy_slow_test.rb]: This test is very funny.

In what case would this fail?

**Comment by kmcphillips** [comment]
Looks great.

File: Rakefile [hunk 4]
```diff
@@ -5,6 +5,13 @@ require "rake/testtask"
 
 Rake::TestTask.new do |task|
   task.libs << "test"
+  task.test_files = Dir["test/**/*_test.rb"] - Dir["test/**/*_slow_test.rb"]
```
Comment (samuelkadolph) [inline Rakefile]: Would `Dir["test/**/*_test.rb"]` not grab the `_slot_test.rb` files as well? And why the need for the 2nd task?

File: lib/buffered_logger.rb (hunk 1 above)
Comment (samuelkadolph) [inline lib/buffered_logger.rb]: We're the only consumers of this gem anyways so :+1:. Plus sweep wasn't really meant to be called manually, just randomly when a logger was ended.

File: lib/buffered_logger/log_device_proxy.rb [hunk 5]
```diff
@@ -12,44 +14,54 @@ def close
...
+      else
+        Thread.current[THREAD_LOCAL_VAR_NAME]
+      end
+    end
+
+    def string_io=(s_io)
```
Comment (samuelkadolph) [inline lib/buffered_logger/log_device_proxy.rb]: I'm not a fan of short form variables. Can we just call it `io`?

File: lib/buffered_logger/log_device_proxy.rb (hunk 2 above)
Comment (sunblaze) [inline lib/buffered_logger/log_device_proxy.rb]: Wanted to use thread local variable instead of fiber locals, don't think I had a use case why. This might be a premature optimization.

Comment (samuelkadolph) [inline lib/buffered_logger/log_device_proxy.rb]: Makes sense. A while back I had the key based on the current Thread and Fiber but realized it didn't make any sense to include the Fiber because Fibers are concurrent. `#thread_variable_get` seems like a good idea, maybe we should make the gem require ruby 2.0 and above and just ditch using `#[]`.

File: test/buffered_logger/log_device_proxy_slow_test.rb (hunk 3 above)
Comment (sunblaze) [inline test/buffered_logger/log_device_proxy_slow_test.rb]: This is a brute force test to recreate the issue I was seeing in production. If you remove the fixes in the PR you'll usually get failures and exceptions when running this test.

File: Rakefile (hunk 4 above)
Comment (sunblaze) [inline Rakefile]: The wildcard should consume `_` and alpha numeric characters so it would normally include the slow tests which I don't want for normal CI runs.

File: lib/buffered_logger.rb (hunk 1 above)
Comment (sunblaze) [inline lib/buffered_logger.rb]: Sounds good.

File: lib/buffered_logger/log_device_proxy.rb (hunk 5 above)
Comment (sunblaze) [inline lib/buffered_logger/log_device_proxy.rb]: Yep

File: lib/buffered_logger/log_device_proxy.rb (hunk 2 above)
Comment (sunblaze) [inline lib/buffered_logger/log_device_proxy.rb]: I'm good with ditching 1.9 :+1:

File: Rakefile (hunk 4 above)
Comment (samuelkadolph) [inline Rakefile]: Oooh, I missed the minus. Thought it was a plus

**Comment by sunblaze** [comment]
@samuelkadolph made those fixes, do you mind checking it out again?

**Comment by samuelkadolph** [comment]
LGTM

**Comment by sunblaze** [comment]
:ship: please :smile:

**Comment by samuelkadolph** [comment]
Released: https://rubygems.org/gems/buffered-logger/versions/2.0.0

**Comment by garethson** [comment]
@samuelkadolph @sunblaze  :100:
